File: backend/app/schemas/batch.py
from datetime import date, datetime
from pydantic import BaseModel, field_validator

from app.models.batch import BatchStatus


class BatchBase(BaseModel):
    drug_id: int
    batch_number: str
    production_date: date | None = None
    expiry_date: date
    quantity: int = 0
    unit: str = "盒"


class BatchCreate(BatchBase):
    @field_validator("batch_number")
    @classmethod
    def batch_number_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("批号不能为空")
        return v.strip()

    @field_validator("quantity")
    @classmethod
    def quantity_non_negative(cls, v: int) -> int:
        if v < 0:
            raise ValueError("初始库存量不能为负")
        return v


class BatchUpdate(BaseModel):
    """部分更新，所有字段可选"""
    batch_number: str | None = None
    production_date: date | None = None
    expiry_date: date | None = None
    unit: str | None = None


class BatchResponse(BaseModel):
    id: int
    drug_id: int
    drug_name: str          # 联查药品名
    batch_number: str
    production_date: date | None
    expiry_date: date
    quantity: int
    unit: str
    status: BatchStatus
    source_ocr_id: int | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class BatchListQuery(BaseModel):
    """批次列表查询参数"""
    drug_id: int | None = None        # 按药品过滤
    status: BatchStatus | None = None  # 按状态过滤
    keyword: str | None = None        # 批号关键词
    page: int = 1
    page_size: int = 20

File: frontend/src/types/batch.ts
export type BatchStatus = 'normal' | 'near_expiry' | 'expired'

export interface Batch {
  id: number
  drug_id: number
  drug_name: string
  batch_number: string
  production_date: string | null
  expiry_date: string
  quantity: number
  unit: string
  status: BatchStatus
  source_ocr_id: number | null
  created_at: string
  updated_at: string
}

export interface BatchCreate {
  drug_id: number
  batch_number: string
  production_date?: string
  expiry_date: string
  quantity: number
  unit?: string
}

export interface BatchUpdate {
  batch_number?: string
  production_date?: string
  expiry_date?: string
  unit?: string
}

export interface BatchListQuery {
  drug_id?: number
  status?: BatchStatus
  keyword?: string
  page?: number
  page_size?: number
}

/** 状态显示映射 */
export const BATCH_STATUS_LABEL: Record<BatchStatus, string> = {
  normal: '正常',
  near_expiry: '临近过期',
  expired: '已过期',
}

export const BATCH_STATUS_TAG: Record<BatchStatus, string> = {
  normal: 'success',
  near_expiry: 'warning',
  expired: 'danger',
}
